cpp/inc/bond/ext/grpc/unary_call.h

```diff
@@ -87,163 +134,282 @@ namespace detail {
-    }
+        /// @brief Get the server context for this call.
+        grpc::ServerContext& context()
+        {
+            return impl()._context;
+        }
 
-    void swap(unary_call& rhs) noexcept
-    {
-        using std::swap;
-        swap(_impl, rhs._impl);
-    }
+        /// @brief Get the request message for this call.
+        const TRequest& request() const
+        {
+            return impl()._request;
+        }
 
-    /// @brief Get the server context for this call.
-    const grpc::ServerContext& context() const
-    {
-        BOOST_ASSERT(_impl);
-        return _impl->_context;
-    }
+        /// @brief Get the request message for this call.
+        TRequest& request()
+        {
+            return impl()._request;
+        }
 
-    /// @brief Get the server context for this call.
-    grpc::ServerContext& context()
-    {
-        BOOST_ASSERT(_impl);
-        return _impl->_context;
-    }
+        /// @brief Responds to the client with the given message and status.
+        ///
+        /// Only the first call to \p Finish or \p FinishWithError will be
+        /// honored.
+        void Finish(const TResponse& msg, const grpc::Status& status = grpc::Status::OK)
+        {
+            Finish(bond::bonded<TResponse>{msg}, status);
+        }
 
-    /// @brief Get the request message for this call.
-    const TRequest& request() const
-    {
-        BOOST_ASSERT(_impl);
-        return _impl->_request;
-    }
+        /// @brief Responds to the client with the given message and status.
+        ///
+        /// Only the first call to \p Finish or \p FinishWithError will be
+        /// honored.
+        void Finish(const bond::bonded<TResponse>& msg, const grpc::Status& status = grpc::Status::OK)
+        {
+            impl().Finish(msg, status);
+        }
 
-    /// @brief Get the request message for this call.
-    TRequest& request()
-    {
-        BOOST_ASSERT(_impl);
-        return _impl->_request;
-    }
+        /// @brief Responds to the client with the given status and no message.
+        ///
+        /// Only the first call to \p Finish or \p FinishWithError will be
+        /// honored.
+        void FinishWithError(const grpc::Status& status)
+        {
+            impl().FinishWithError(status);
+        }
+
+    protected:
+        explicit unary_call_base(impl_type*&& impl) noexcept
+            : _impl(std::move(impl))
+        {
+            BOOST_ASSERT(_impl);
+        }
+
+        explicit operator bool() const
+        {
+            return static_cast<bool>(_impl);
+        }
+
+    private:
+        boost::intrusive_ptr<impl_type> _impl;
+
+        impl_type& impl()
+        {
+            BOOST_ASSERT(_impl);
+            return *_impl;
+        }
+
+        const impl_type& impl() const
+        {
+            BOOST_ASSERT(_impl);
+            return *_impl;
+        }
+    };
+
+} // namespace detail
+
+template <typename TRequest, typename TResponse>
+class shared_unary_call;
```

Comment:
Please separate each class into its own header.

Reply:
Will do after resolving the bulk of the other comments, as GitHub has trouble tracking comments across moves.